### PR TITLE
Implement timeouts for both sending traces and telemetry.

### PR DESCRIPTION
**Pull request by hoolioh**
# What does this PR do?

This PR adds a new timeout field to the `Endpoint` struct in order to set a timeout for outgoing connections used for sending telemetry and traces.

# Motivation

Prevent situations where holding too many connections could lead to memory pressure or resource exhaustion in the system by dropping them after a period of inactivity.

### CODE REVIEW

**Comment by bwoebi** [comment]
Slightly surprised that we opt modifying the Endpoint struct, but it probably makes sense :-)
 
Can you please also add a FFI API to set the timeout on the Endpoint?

File: ddcommon/src/lib.rs [hunk 1]
```diff
@@ -32,11 +32,22 @@ pub type HttpClient = hyper::Client<connector::Connector, hyper::Body>;
 pub type HttpResponse = hyper::Response<hyper::Body>;
 pub type HttpRequestBuilder = hyper::http::request::Builder;
 
-#[derive(Default, Clone, PartialEq, Eq, Hash, Debug, Serialize, Deserialize)]
+#[derive(Clone, PartialEq, Eq, Hash, Debug, Serialize, Deserialize)]
 pub struct Endpoint {
     #[serde(serialize_with = "serialize_uri", deserialize_with = "deserialize_uri")]
     pub url: hyper::Uri,
     pub api_key: Option<Cow<'static, str>>,
+    pub timeout: u64,
```
Comment (bantonsson) [inline ddcommon/src/lib.rs:40]: Is there a particular reason to not have this as a `Duration`?

Comment (hoolioh) [reply ddcommon/src/lib.rs:40]: The aim was to reduce a little bit of memory. Since `Endpoint` is embedded in each `SendData` structurure and, therefore, the `TraceFlusher` holds a collection of `SendData` objects I thought it would be better just to hold 8 bytes rather than 16 per `SendData` instance. I even thought of just holding a u16 (probably a timeout of 65535 would be enough for our purposes) but since all milliseconds values inside std::time were expressed as a u64 I left it that way.

File: bin_tests/src/bin/crashtracker_bin_test.rs [hunk 2]
```diff
@@ -46,6 +46,7 @@ mod unix {
             Some(ddcommon::Endpoint {
                 url: ddcommon::parse_uri(&output_url)?,
                 api_key: None,
```
Comment (bantonsson) [inline bin_tests/src/bin/crashtracker_bin_test.rs:48]: I guess that all occurrences of these could be removed.

**Review by bantonsson — APPROVED** [review]

File: bin_tests/src/bin/crashtracker_bin_test.rs (hunk 2 above)
Comment (hoolioh) [reply bin_tests/src/bin/crashtracker_bin_test.rs:48]: you're right. I missed some of them.

**Review by pawelchcki — APPROVED** [review]
Generally looks good to me. I really liked the async mocked http tests !

Great addition thanks Julio! :bow: 

Added one optional suggestion taking a look at `thiserror` :) 

Though more importantly - when we plan to connect through world wide internet. 
We should consider giving our connector a bit more smartness with regards to timeout granularity.
e.g. copy impl from https://github.com/hjr3/hyper-timeout/blob/master/src/lib.rs

Basically in this PR we add more less one coarse timeout. 

And as I had to debug bugs in glibc sourced DNS resolution lock ups in ruby some time ago, I'm pretty paranoid. 

In ideal world I would love to see:
1. separate timeout and caching for dns resolution (esp if we're connecting to wide internet)
2. timeout on connect
3. separate timeouts on reads and writes.

This could give our users more visibility into what is happening in the background. 

^ just some thoughts for the future.

This PR -> LGTM!

File: trace-utils/src/send_data/mod.rs [hunk 3]
```diff
@@ -853,17 +889,119 @@ mod tests {
...
+                timeout: MILLIS_PER_SECOND,
             },
         );
 
         let res = data.send().await;
 
         assert!(res.last_result.is_err());
-        assert_eq!(res.errors_timeout, 0);
-        assert_eq!(res.errors_network, 1);
+        match std::env::consts::OS {
```
Comment (VianneyRuhlmann) [inline trace-utils/src/send_data/mod.rs:899]: Why do we expect a timeout for windows instead of a network error ?

Comment (hoolioh) [reply trace-utils/src/send_data/mod.rs:899]: Oh, that drove me crazy. I'm not really familiar with Windows TCP/IP stack but it seems that, at hyper level, it returns a timeout rather than a connection refused error despite not having a listening socket on that port (4321).

**Comment by danielsn** [comment]
There are several APIs that currently take timeouts and endpoints in their API.  e.g.
https://github.com/DataDog/libdatadog/blob/main/profiling/src/exporter/mod.rs#L173
https://github.com/DataDog/libdatadog/blob/main/crashtracker/src/configuration.rs#L29

These should be updated to only take the new endpoint

File: ddcommon-ffi/src/endpoint.rs [hunk 4]
```diff
@@ -10,9 +10,12 @@ use std::str::FromStr;
...
 #[must_use]
 pub extern "C" fn ddog_endpoint_from_url(url: crate::CharSlice) -> Option<Box<Endpoint>> {
-    parse_uri(url.to_utf8_lossy().as_ref())
-        .ok()
-        .map(|url| Box::new(Endpoint { url, api_key: None }))
+    parse_uri(url.to_utf8_lossy().as_ref()).ok().map(|url| {
+        Box::new(Endpoint {
+            url,
+            ..Default::default()
+        })
```
Comment (danielsn) [inline ddcommon-ffi/src/endpoint.rs:17]: This pattern happens enough we should just have a constructor `Endpoint::from_url()`

File: ddcommon/src/lib.rs (hunk 1 above)
Comment (danielsn) [reply ddcommon/src/lib.rs:40]: At the very least, this should document the unit, e.g. `timeout_ms`.  I'd argue for the `Duration` type as more self-documenting: do we have numbers for how many bytes this actually saves?

File: ddcommon/src/lib.rs [hunk 5]
```diff
@@ -133,6 +144,9 @@ pub fn decode_uri_path_in_authority(uri: &hyper::Uri) -> anyhow::Result<PathBuf>
 }
 
 impl Endpoint {
+    /// Default value for the timeout field in milliseconds.
+    pub const DEFAULT_TIMEOUT: u64 = 10_000;
```
Comment (danielsn) [inline ddcommon/src/lib.rs:148]: how did we come up with this number?

File: ddtelemetry/examples/tm-send-sketch.rs [hunk 6]
```diff
@@ -116,6 +116,7 @@ async fn async_main() {
             "https://instrumentation-telemetry-intake.datad0g.com/api/v2/apmtelemetry",
         ),
         api_key: Some(Cow::Owned(std::env::var("DD_API_KEY").unwrap())),
+        ..Default::default()
```
Comment (danielsn) [inline ddtelemetry/examples/tm-send-sketch.rs:119]: why is this needed?

File: ddtelemetry/src/config.rs [hunk 7]
```diff
@@ -228,7 +228,11 @@ impl Config {
             restartable: false,
         };
         if let Ok(url) = parse_uri(&trace_agent_url) {
-            let _res = this.set_endpoint(Endpoint { url, api_key });
+            let _res = this.set_endpoint(Endpoint {
+                url,
+                api_key,
+                ..Default::default()
```
Comment (danielsn) [inline ddtelemetry/src/config.rs:234]: why is this needed?

File: ddtelemetry/src/worker/mod.rs [hunk 8]
```diff
@@ -183,6 +183,9 @@ mod serialize {
 }
 
 impl TelemetryWorker {
+    // Default request timeout in milliseconds.
+    const DEFAULT_TIMEOUT: u64 = 10_000;
```
Comment (danielsn) [inline ddtelemetry/src/worker/mod.rs:187]: This could be imported from the Endpoint timeout, avoid duplicating the same constant

File: profiling/src/exporter/config.rs [hunk 9]
```diff
@@ -28,7 +28,10 @@ pub fn agent(base_url: Uri) -> anyhow::Result<Endpoint> {
     };
     parts.path_and_query = p_q;
     let url = Uri::from_parts(parts)?;
-    Ok(Endpoint { url, api_key: None })
+    Ok(Endpoint {
+        url,
+        ..Default::default()
+    })
```
Comment (danielsn) [inline profiling/src/exporter/config.rs:34]: As before, would be clearer as `Endpoint::from_url(...)`

**Review by danielsn — DISMISSED** [review]

File: trace-utils/src/send_data/mod.rs [hunk 10]
```diff
@@ -458,6 +471,7 @@ mod tests {
     use httpmock::MockServer;
     use std::collections::HashMap;
 
+    const MILLIS_PER_SECOND: u64 = 1_000;
```
Comment (ekump) [inline trace-utils/src/send_data/mod.rs:474]: I'm not sure this is the best description of the value (even though it is technically true). Maybe `TIMEOUT_MS` is more correct?

File: sidecar/src/config.rs [hunk 11]
```diff
@@ -182,6 +182,7 @@ pub fn get_product_endpoint(subdomain: &str, endpoint: &Endpoint) -> Endpoint {
         Endpoint {
             url: hyper::Uri::from_parts(parts).unwrap(),
             api_key: Some(api_key.clone()),
+            ..Default::default()
```
Comment (ekump) [inline sidecar/src/config.rs:185]: This also applies to the data-pipeline implementation...Is the current default of 10s appropriate? And do we want the same timeout value for both agent and agentless communication? I would think we would want a much lower value for talking to a trace agent? I'm not sure what appropriate values are, maybe we could check with trace agent folks?

File: ddcommon/src/lib.rs (hunk 1 above)
Comment (hoolioh) [reply ddcommon/src/lib.rs:40]: @danielsn: Yeah, that'll make things clearer. Didn't have time to do benchnmark it. 
@ekump: `SendData` instances are eventually merged together (up to a maximum payload size) when the `TraceFlusher` tries flush them out so not every timeout is converted to a `Duration` that's why it seemed a waste of space.

File: trace-utils/src/send_data/mod.rs (hunk 3 above)
Comment (VianneyRuhlmann) [reply trace-utils/src/send_data/mod.rs:899]: Why am I not surprised 😩. Maybe add a comment to avoid confusion

File: data-pipeline/src/trace_exporter.rs [hunk 12]
```diff
@@ -310,21 +310,17 @@ impl TraceExporterBuilder {
...
             .enable_all()
             .build()?;
 
         Ok(TraceExporter {
-            endpoint,
+            endpoint: Endpoint::from_slice(
+                self.url
+                    .to_owned()
+                    .unwrap_or("http://127.0.0.1:8126/".to_string())
+                    .as_str(),
```
Comment (danielsn) [inline data-pipeline/src/trace_exporter.rs:322]:
```suggestion
                self.url
                    .as_ref()
                    .map(|s| s.as_str())
                    .unwrap_or("http://127.0.0.1:8126/"),
```

File: sidecar/src/dogstatsd.rs [hunk 13]
```diff
@@ -185,7 +185,7 @@ mod test {
                 .as_str()
                 .parse::<Uri>()
                 .unwrap(),
-            api_key: None,
+            ..Default::default()
```
Comment (danielsn) [inline sidecar/src/dogstatsd.rs:188]:
```
Endpoint::from_slice(socket
                .local_addr()
                .unwrap()
                .to_string()
                .as_str())
```

File: sidecar/src/dogstatsd.rs [hunk 14]
```diff
@@ -250,15 +241,14 @@ mod test {
     fn test_create_client_unix_domain_socket() {
         let res = create_client(Some(Endpoint {
             url: "unix://localhost:80".parse::<Uri>().unwrap(),
-            api_key: None,
+            ..Default::default()
```
Comment (danielsn) [inline sidecar/src/dogstatsd.rs:244]: `Endpoint::from_slice`

File: sidecar/src/config.rs (hunk 11 above)
Comment (bwoebi) [reply sidecar/src/config.rs:181]: This code should inherit the endpoint from input endpoint var.

Comment (bwoebi) [reply sidecar/src/config.rs:181]: I think it's the responsibility of the code setting the agentless mode to also decide on the timeout. But yes, I have no idea what proper timeouts are. Probably `Endpoint::from_url()` may default to a lower 3 sec timeout?

**Review by bwoebi — APPROVED** [review]
Looks good to me now, thanks!

File: bin_tests/src/bin/crashtracker_bin_test.rs [hunk 15]
```diff
@@ -37,15 +37,15 @@ mod unix {
...
         let wait_for_receiver = true;
 
         let endpoint = if output_url.is_empty() {
             None
         } else {
-            Some(ddcommon::Endpoint {
-                url: ddcommon::parse_uri(&output_url)?,
-                api_key: None,
+            Some(Endpoint {
+                url: parse_uri(&output_url).unwrap(),
```
Comment (danielsn) [inline bin_tests/src/bin/crashtracker_bin_test.rs:46]: `Endpoint::from_slice`

File: crashtracker/src/api.rs [hunk 16]
```diff
@@ -142,8 +141,9 @@ fn test_crash() -> anyhow::Result<()> {
     let output_url = format!("file://{dir}{time}.txt");
 
     let endpoint = Some(Endpoint {
-        url: parse_uri(&output_url).unwrap(),
-        api_key: None,
+        url: ddcommon::parse_uri(&output_url).unwrap(),
```
Comment (danielsn) [inline crashtracker/src/api.rs:144]: `Endpoint::from_slice`

File: data-pipeline/src/trace_exporter.rs [hunk 17]
```diff
@@ -310,21 +310,12 @@ impl TraceExporterBuilder {
...
-            )?,
-            api_key: None,
-        };
         let runtime = tokio::runtime::Builder::new_current_thread()
             .enable_all()
             .build()?;
 
         Ok(TraceExporter {
-            endpoint,
+            endpoint: Endpoint::from_slice(self.url.as_deref().unwrap_or("http://127.0.0.1:8126")),
```
Comment (danielsn) [inline data-pipeline/src/trace_exporter.rs:318]: 💯

File: examples/ffi/exporter.cpp [hunk 18]
```diff
@@ -140,10 +140,11 @@ int main(int argc, char *argv[]) {
   ddog_CharSlice info_example = DDOG_CHARSLICE_C_BARE(
       "{\"application\": {\"start_time\": \"2024-01-24T11:17:22+0000\"}, \"platform\": {\"kernel\": \"Darwin Kernel 22.5.0\"}}");
 
+  ddog_prof_Exporter_set_timeout(exporter, 30000);
```
Comment (danielsn) [inline examples/ffi/exporter.cpp:143]: I wonder if we even want this?  Tho as an example, maybe its good to show off the new API

File: profiling-ffi/src/exporter.rs [hunk 19]
```diff
@@ -192,6 +192,20 @@ fn ddog_prof_exporter_new_impl(
...
+/// Sets the timeout for the exporte's timeout.
+/// # Arguments
+/// * `exporter` - ProfileExporter instance.
+/// * `timeout_ms` - timeout in milliseconds.
+#[no_mangle]
+pub unsafe extern "C" fn ddog_prof_Exporter_set_timeout(
+    exporter: Option<&mut ProfileExporter>,
+    timeout_ms: u64,
+) {
+    if let Some(ptr) = exporter {
```
Comment (danielsn) [inline profiling-ffi/src/exporter.rs:204]: Should we return an error in this case?

File: profiling/tests/form.rs [hunk 20]
```diff
@@ -35,7 +35,8 @@ fn multipart(
     let start = now.sub(chrono::Duration::seconds(60));
     let end = now;
 
-    let timeout = std::time::Duration::from_secs(10);
+    let timeout: u64 = 10;
```
Comment (danielsn) [inline profiling/tests/form.rs:38]: This turns a 10 **second** timeout to 10 **ms**

File: trace-utils/src/send_data/mod.rs [hunk 21]
```diff
@@ -977,7 +1120,7 @@ mod tests {
 
         let target_endpoint = Endpoint {
             url: server.url("").to_owned().parse().unwrap(),
-            api_key: None,
+            ..Default::default()
```
Comment (danielsn) [inline trace-utils/src/send_data/mod.rs:1123]: `Endpoint::from_slice("")`

**Review by danielsn — APPROVED** [review]
Good to go, modulo some small comments.
